### PR TITLE
TSK-1668: Survey plugin

### PR DESCRIPTION
**Pull request by anotherpit**
# Contribution checklist

## Brief description

Might be mergeable, but still not a complete feature, Work In Progress. Let's say it's a Developer Preview for early comments.

## Checklist

* [ ] - UI test added to added/changed functionality?
* [ ] - Screenshot is added to PR if applicable ?
* [ ] - Does a local formatting is applied (rush format)
* [ ] - Does a local svele-check is applied (rush svelte check)
* [ ] - Does a local UI tests are executed [UI Testing](../tests/readme.md)
* [ ] - Does the code work? Check whether function and logic are correct.
* [ ] - Does Changelog.md is updated with changes?
* [ ] - Does the translations are up to date?
* [ ] - Does it well tested?
* [ ] - Tested for Chrome.
* [ ] - Tested for Safari.
* [ ] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [ ] - Rebase your branch onto master and upstream branch
* [ ] - Is there any redundant or duplicate code?
* [ ] - Are required links are linked to PR?
* [ ] - Does new code is well documented ?

## Related issues

A list of closed updated issues

## Contributor requirements

* [x] - Sign-off is provided. [DCO](https://github.com/apps/dco)
* [x] - GPG Signature is provided. [GPG](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)

### CODE REVIEW

File: packages/core/src/classes.ts [hunk 1]
```diff
@@ -41,6 +41,11 @@ export type Markup = string
  */
 export type Hyperlink = string
 
+/**
+ * @public
+ */
+export type CollectionSize<T> = T[]['length']
```
Comment (anotherpit) [inline packages/core/src/classes.ts:47]: Slightly more expressive way to declare collection fields. Transpiles to good old `number`, but is more self-explanatory when reading code

File: plugins/survey/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ * Assessable interactive survey item, e.g. a question with correct answers
+ */
+export interface SurveyAssessableItem extends SurveyInteractiveItem {
+  weight: Percentage
+}
+
+/**
+ * @public
+ */
+export interface Survey extends AttachedDoc {
```
Comment (anotherpit) [inline plugins/survey/src/index.ts:63]: Model interfaces are gonna change significantly in further commits/PRs. Feel free to ignore them for now

File: plugins/tracker/src/index.ts [hunk 3]
```diff
@@ -185,7 +186,7 @@ export interface Issue extends Task {
   // ReportedTime time, auto updated using trigger.
   reportedTime: number
   // Collection of reportedTime entries, for proper time estimations per person.
-  reports: number
+  reports: CollectionSize<TimeSpendReport>
```
Comment (anotherpit) [inline plugins/tracker/src/index.ts:189]: Example of usage of new `CollectionSize` helper type

File: models/survey/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import core from '@hcengineering/model-core'
+import workbench from '@hcengineering/model-workbench'
+import { surveyId } from '@hcengineering/survey'
+import survey from './plugin'
+
+export { surveyId } from '@hcengineering/survey'
+export { default } from './plugin'
+
+export function createModel (builder: Builder): void {
+  builder.createDoc(
```
Comment (anotherpit) [inline models/survey/src/index.ts:27]: Not sure if we want to unconditionally register an application for `Survey`. Or, at least, if we want to do it in Platform. I kinda need it for dev/debug needs, but I can safely keep it out of `main` branch

File: models/all/src/index.ts [hunk 5]
```diff
@@ -260,6 +261,16 @@ export default function buildModel (enabled: string[] = ['*'], disabled: string[
       }
     ],
     [supportModel, supportId],
+    [
+      surveyModel,
+      surveyId,
+      {
+        // TODO: icon, label, description?
```
Comment (anotherpit) [inline models/all/src/index.ts:270]: What are these props for? They do not affect the application entry in sidebar (which is controlled by its own `Application` model instance), so why are they?

File: packages/core/src/classes.ts (hunk 1 above)
Comment (aonnikov) [reply packages/core/src/classes.ts:47]: nice! I would do the same for "sequence" numbers, counters that are used for safe sequential numbers generation

File: models/survey/src/index.ts (hunk 4 above)
Comment (aonnikov) [reply models/survey/src/index.ts:27]: That's ok. If needed, we can disable it for deployment.

File: models/all/src/index.ts (hunk 5 above)
Comment (aonnikov) [reply models/all/src/index.ts:270]: I think these supposed to be used to show app name and icon in settings / configuration

File: plugins/survey/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  rank: Rank
+}
+
+/**
+ * @public
+ *
+ * Interactive survey item, e.g. a question
+ */
+export interface SurveyInteractiveItem extends SurveyItem {
+  // TODO: Or plain string?
```
Comment (aonnikov) [inline plugins/survey/src/index.ts]: I would use plain string

File: models/all/src/index.ts (hunk 5 above)
Comment (haiodo) [reply models/all/src/index.ts:270]: Yes, this is for configuration, so entire application could be turned on/off per workspace.

File: templates/ui/package.json [hunk 7]
```diff
@@ -8,11 +8,11 @@
...
     "format": "format src"
   },
-  "devDependencies": {    
+  "devDependencies": {
     "@hcengineering/platform-rig": "^0.6.0",
-    "svelte-loader":"^3.1.3",
+    "svelte-loader":"^3.1.9",
     "sass":"^1.53.0",
-    "svelte-preprocess":"5.0.3",
+    "svelte-preprocess":"^5.1.0",
```
Comment (haiodo) [inline templates/ui/package.json:15]: You need to apply to all plugins after changing this file, so let's make it as separate PR.

Comment (anotherpit) [reply templates/ui/package.json:15]: I've just checked. _All_ plugins do already depend on `^5.1.0`

File: plugins/survey/src/index.ts (hunk 6 above)
Comment (anotherpit) [reply plugins/survey/src/index.ts]: ✅

File: templates/ui/package.json (hunk 7 above)
Comment (anotherpit) [reply templates/ui/package.json:15]: ✅

File: models/all/src/index.ts (hunk 5 above)
Comment (anotherpit) [reply models/all/src/index.ts:270]: ✅

File: models/survey/src/index.ts (hunk 4 above)
Comment (anotherpit) [reply models/survey/src/index.ts:27]: ✅

File: plugins/tracker/src/index.ts (hunk 3 above)
Comment (anotherpit) [reply plugins/tracker/src/index.ts:189]: ✅

File: plugins/survey/src/index.ts (hunk 2 above)
Comment (anotherpit) [reply plugins/survey/src/index.ts:63]: ✅

**Review by aonnikov — DISMISSED** [review]

**Review by aonnikov — APPROVED** [review]
